forem/forem
# add field for billboard target geolocations, with unit and e2e tests

## What type of PR is this? (check all applicable)


## Description

Following up on https://github.com/forem/forem/pull/19834, this change adds a field during the creation and updating of a Billboard, and lets users set comma-separated ISO3166-2 geolocation codes (US and Canada only) according to the validations already present on the model.

This change didn't have to be too large because the model is already good at displaying specific error messages for its failed validations. We're just adding a small helper method to ensure the typed input persists in its comma-separated format as the user resubmits and the page refreshes.

## Related Tickets & Documents



- Related Issue #19794
- Closes #19794

## QA Instructions, Screenshots, Recordings

The feature flag `billboard_location_targeting` needs to be enabled to test this functionality. In localhost, I did this by going to http://localhost:3000/admin/feature_flags/features and turning it all the way on, and I believe the Admin test account will be able to do a similar toggling in Uffizzi once I draft this PR.

With the flag enabled, whether creating or editing a billboard ("Admin -> Customization -> Display Ads"), you should now see a "Target Geolocations: field with placeholder examples of the accepted format and the countries we currently support.

Errors will generate from that field if you 
  attempt to create/update with an unsupported country's valid code e.g. "MX-CMX" or
  attempt to create/update with a supported country and invalid region e.g. "US-UHHHH"
  
With otherwise valid inputs, a string of comma separated valid inputs for geolocations e.g. "US-NV, US-TX, CA-BC, CA-NS" should allow a successful save and persist on that display ad when its Show page is loaded.

If some inputs are valid and others are invalid, only the invalid ones should show up in the flash notice about form errors.

### UI accessibility checklist
_If your PR includes UI changes, please utilize this checklist:_
- [X] Keyboard operability supported? Unrelated note to self: tabbing to the "Display to user group" and "Type" radio buttons works, but there's no visual indicator that the radio groups are currently tabbed to like there is with outlined text fields. Will see if there's already an issue for this.

_For more info, check out the
[Forem Accessibility Docs](https://developers.forem.com/frontend/accessibility)._

## Added/updated tests?

      have not been included_

## [optional] Are there any post deployment tasks we need to perform?

Make sure the flipper feature is enabled for the actors who plan to test it

## [optional] What gif best describes this PR or how it makes you feel?

![Gina Rodriguez from the Netflix Carmen Sandiego saying "I wanted you to find me"](https://media.tenor.com/JSxQlfeP5WsAAAAC/i-wanted-you-to-find-me-gina-rodriguez.gif)
